velox/functions/lib/DateTimeImpl.h

```diff
@@ -0,0 +1,204 @@
+              .count() /
+          7;
+    }
+    default:
+      break;
+  }
+
+  // Month, quarter and year do not have fixed conversion ratio. Ex. a month can
+  // have 28, 29, 30 or 31 days. A year can have 365 or 366 days.
+  const std::chrono::time_point<std::chrono::system_clock, date::days>
+      fromDaysTimepoint = std::chrono::floor<date::days>(fromTimepoint);
+  const std::chrono::time_point<std::chrono::system_clock, date::days>
+      toDaysTimepoint = std::chrono::floor<date::days>(toTimepoint);
+  const date::year_month_day fromCalDate(fromDaysTimepoint);
+  const date::year_month_day toCalDate(toDaysTimepoint);
+  const uint64_t fromTimeInstantOfDay =
+      std::chrono::duration_cast<std::chrono::milliseconds>(
+          fromTimepoint - fromDaysTimepoint)
+          .count();
+
+  uint64_t toTimeInstantOfDay = 0;
+  uint64_t toTimePointMillis = toTimepoint.time_since_epoch().count();
+  uint64_t toDaysTimepointMillis =
+      std::chrono::
+          time_point<std::chrono::system_clock, std::chrono::milliseconds>(
+              toDaysTimepoint)
+              .time_since_epoch()
+              .count();
+  bool overflow = __builtin_sub_overflow(
+      toTimePointMillis, toDaysTimepointMillis, &toTimeInstantOfDay);
+  VELOX_USER_CHECK_EQ(
+      overflow,
+      false,
+      "{} - {} Causes arithmetic overflow: {} - {}",
+      fromTimestamp.toString(),
+      toTimestamp.toString(),
+      toTimePointMillis,
+      toDaysTimepointMillis);
+  const uint8_t fromDay = static_cast<unsigned>(fromCalDate.day()),
+                fromMonth = static_cast<unsigned>(fromCalDate.month());
+  const uint8_t toDay = static_cast<unsigned>(toCalDate.day()),
+                toMonth = static_cast<unsigned>(toCalDate.month());
+  const date::year_month_day toCalLastYearMonthDay(
+      toCalDate.year() / toCalDate.month() / date::last);
+  const uint8_t toLastYearMonthDay =
+      static_cast<unsigned>(toCalLastYearMonthDay.day());
+
+  if (unit == DateTimeUnit::kMonth || unit == DateTimeUnit::kQuarter) {
+    int64_t diff =
+        (int64_t(toCalDate.year()) - int64_t(fromCalDate.year())) * 12 +
+        int(toMonth) - int(fromMonth);
+
+    if (((!respectLastDay || toDay != toLastYearMonthDay) && fromDay > toDay) ||
+        (fromDay == toDay && fromTimeInstantOfDay > toTimeInstantOfDay)) {
+      diff--;
+    }
+
+    diff = (unit == DateTimeUnit::kMonth) ? diff : diff / 3;
+    return sign * diff;
+  }
+
+  if (unit == DateTimeUnit::kYear) {
+    int64_t diff = (toCalDate.year() - fromCalDate.year()).count();
+
+    if (fromMonth > toMonth ||
+        (fromMonth == toMonth && fromDay > toDay &&
+         (!respectLastDay || toDay != toLastYearMonthDay)) ||
+        (fromMonth == toMonth && fromDay == toDay &&
+         fromTimeInstantOfDay > toTimeInstantOfDay)) {
+      diff--;
+    }
+    return sign * diff;
+  }
+
+  VELOX_UNREACHABLE("Unsupported datetime unit");
```

Comment:
> * Use `VELOX_UNREACHABLE()` when a particular branch/block should never be
  executed, such as in a switch statement with an invalid `default:` block.
> * Use `VELOX_NYI()` for features or code paths that are not implemented yet. (throws runtime error)

`VELOX_UNSUPPORTED` throws user error for unsupported features.

Which is the case here?

Reply:
The first one, all possible units have been processed, similar to default: block.

Reply:
If so, you could just call `VELOX_UNREACHABLE()` without message.